File: packages/lynx-audio/android/com/sigx/audio/AudioPublisher.kt
package com.sigx.audio

import android.util.Log
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.tasm.LynxView
import java.util.UUID

/**
 * Per-[LynxView] publisher that pumps [AudioEventBus] payloads into JS
 * via `LynxView.sendGlobalEvent(channel, [...])`. One instance per
 * LynxView; the autolinker generates `GeneratedLifecyclePublishers` which
 * instantiates and attaches us for each view.
 */
class AudioPublisher(private val lynxView: LynxView) {

    private var token: UUID? = null

    fun attach() {
        token = AudioEventBus.addListener { channel, payload ->
            try {
                val params = JavaOnlyArray()
                params.pushMap(payload)
                lynxView.sendGlobalEvent(channel, params)
            } catch (e: Throwable) {
                Log.w(TAG, "publish failed: ${e.message}")
            }
        }
    }

    fun detach() {
        token?.let { AudioEventBus.removeListener(it) }
        token = null
    }

    private companion object {
        const val TAG = "SigxAudioPublisher"
    }
}

File: packages/lynx-audio/ios/AudioPublisher.swift
import Foundation
import Lynx

/// Per-`LynxView` publisher that pumps `AudioEventBus` payloads into JS via
/// `LynxView.sendGlobalEvent(channel, withParams:)`. One instance per
/// LynxView; instantiated by the autolink-generated
/// `GeneratedLifecyclePublishers.attachAll(to:)` and retained for the
/// LynxView's lifetime.
final class AudioPublisher {

    private weak var lynxView: LynxView?
    private var token: UUID?

    init(lynxView: LynxView) {
        self.lynxView = lynxView
        self.token = AudioEventBus.shared.addListener { [weak self] channel, payload in
            guard let view = self?.lynxView else { return }
            view.sendGlobalEvent(channel, withParams: [payload])
        }
    }

    deinit {
        if let token = token {
            AudioEventBus.shared.removeListener(token)
        }
    }
}
